subsys/bluetooth/audio/vocs.c

```diff
@@ -0,0 +1,482 @@
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &inst->state, sizeof(inst->state));
+}
+
+static void location_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	BT_DBG("value 0x%04x", value);
+}
+
+#endif /* CONFIG_BT_VOCS */
+
+static ssize_t write_location(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			      const void *buf, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	struct vocs_server *inst = attr->user_data;
+	uint32_t old_location = inst->location;
+
+	if (len != sizeof(inst->location)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	memcpy(&inst->location, buf, len);
+	BT_DBG("%02x", inst->location);
+
+	if (old_location != inst->location) {
+		bt_gatt_notify_uuid(NULL, BT_UUID_VOCS_LOCATION, inst->service_p->attrs,
+				    &inst->location, sizeof(inst->location));
+
+		if (inst->cb && inst->cb->location) {
+			inst->cb->location(NULL, (struct bt_vocs *)inst, 0, inst->location);
+		}
+	}
+
+	return len;
+}
+
+#if defined(CONFIG_BT_VOCS)
+static ssize_t read_location(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			     void *buf, uint16_t len, uint16_t offset)
+{
+	struct vocs_server *inst = attr->user_data;
+
+	BT_DBG("0x%02x", inst->location);
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &inst->location,
+				 sizeof(inst->location));
+}
+#endif /* CONFIG_BT_VOCS */
+
+static ssize_t write_vocs_control(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				  const void *buf, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	struct vocs_server *inst = attr->user_data;
+	const struct vocs_control_t *cp = buf;
+	bool notify = false;
+
+	if (!len || !buf) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	/* Check opcode before length */
+	if (!VALID_VOCS_OPCODE(cp->opcode)) {
+		BT_DBG("Invalid opcode %u", cp->opcode);
+		return BT_GATT_ERR(BT_VOCS_ERR_OP_NOT_SUPPORTED);
+	}
+
+	if (offset) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (len != sizeof(struct vocs_control_t)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	BT_DBG("Opcode %u, counter %u", cp->opcode, cp->counter);
+
+
+	if (cp->counter != inst->state.change_counter) {
+		return BT_GATT_ERR(BT_VOCS_ERR_INVALID_COUNTER);
+	}
+
+	switch (cp->opcode) {
+	case VOCS_OPCODE_SET_OFFSET:
+		BT_DBG("Set offset %d", cp->offset);
+		if (cp->offset > BT_VOCS_MAX_OFFSET || cp->offset < BT_VOCS_MIN_OFFSET) {
+			return BT_GATT_ERR(BT_VOCS_ERR_OUT_OF_RANGE);
+		}
+
+		if (inst->state.offset != sys_le16_to_cpu(cp->offset)) {
+			inst->state.offset = sys_le16_to_cpu(cp->offset);
+			notify = true;
+		}
+		break;
+	default:
+		return BT_GATT_ERR(BT_VOCS_ERR_OP_NOT_SUPPORTED);
+	}
+
+	if (notify) {
+		inst->state.change_counter++;
+		BT_DBG("New state: offset %d, counter %u",
+		       inst->state.offset, inst->state.change_counter);
+		bt_gatt_notify_uuid(NULL, BT_UUID_VOCS_STATE, inst->service_p->attrs,
```

Comment:
Should there be any handling of the return value here?

Reply:
See https://github.com/zephyrproject-rtos/zephyr/pull/31893#discussion_r598755623